ENH add ResourceTracker to improve ressource clean up

pierreglaser opened the pull request:
This PR aims at extending the current `SemaphoreTracker` to make it able to track more generic named resources, such as folders, and potentially `memmap` and shared memory.

Many small changes in files because of the name refactoring. The main change is the registering/unregistering API extension: when one wants to track a new resource, one must (along with the resource name), specify the resource type. This is necessary for the `RessourceTracker` to know which cleanup function to use for each resource before shutdown.

Base branch: https://github.com/tomMoral/loky/pull/202

ogrisel commented:
French typo everywhere: ressource => resource :)

ogrisel commented on tests/test_resource_tracker.py:
```diff
@@ -0,0 +1,227 @@
+                    the_warn.message)
+            else:
+                print(all_warn)
+                assert len(all_warn) == 0
```
We should get rid of debug prints in test. The `-l` flag of pytest should be informative enough in case of assertion failure.

pierreglaser commented:
I need somehow rebase with #202.

tomMoral commented on loky/backend/resource_tracker.py:
```diff
@@ -0,0 +1,277 @@
+                # process will out live us, so no need to wait on pid
+                exe = spawn.get_executable()
+                args = [exe] + util._args_from_interpreter_flags()
+                # In python 3.3, there is a bug which put `-RRRRR..` instead of
```
This could be removed as we do not support `python3.3` anymore.

pierreglaser replied:
Sure. These lines predated this PR though.

tomMoral commented on loky/backend/resource_tracker.py:
```diff
@@ -0,0 +1,277 @@
+
+
+def main(fd, verbose=0):
+    '''Run semaphore tracker.'''
```
```suggestion
    '''Run resource tracker.'''
```

tomMoral commented on loky/backend/resource_tracker.py:
```diff
@@ -0,0 +1,277 @@
+    signal.signal(signal.SIGTERM, signal.SIG_IGN)
+
+    if _HAVE_SIGMASK:
+        signal.pthread_sigmask(signal.SIG_UNBLOCK, _IGNORED_SIGNALS)
```
I don't understand why we use both `signal.signal` and the `pthread_sigmask`. This feels redundant to me.. Could you please explain why we need both? Or if we do not, let's stick to `signal.signal` as it works for both `win32` and `unix`.

pierreglaser replied:
These lines come from a `multiprocessing` patch. It is already introduced in #202. I did not question them, and simply added them for upstream synchronization concerns.

The point of these lines is to avoid race conditions where `SIGINT` or `SIGTERM` signals are sent to the `semaphore_tracker` process before its custom signal handlers are registered in `main`.
`signal.pthread_sigmask` sets a signal mask in the parent process. The [linux docs](http://man7.org/linux/man-pages/man2/sigprocmask.2.html) tells us that a child process created by `fork` inherits the signal mask of its parent. Temporarily setting a mask before the `semaphore_tracker` is launched was the proposed solution to avoid the race condition described above. Note that signal handling modifications via `signal.signal` are not inheritable. So using `signal.signal` instead of `signal.pthread_mask` in these lines is not possible.

Reciprocally, we cannot rely only on `signal.pthread_mask` because this method is not available in `python2` and on `windows`. Would we only support `unix` systems on `python3`, we could simply keep the `_IGNORED_SIGNALS` blocked and suppress those lines AFAICT.

tomMoral commented on loky/backend/resource_tracker.py:
```diff
@@ -0,0 +1,277 @@
+            try:
+                warnings.warn('resource_tracker: There appear to be %d '
+                              'leaked resources to clean up at shutdown' %
+                              len(cache))
```
I think that it would be better to let the user know what type of resource he is leaking, so he can fix the leak :)

For this, I would use a dict with a set per `rtype` for the `cache`:
```
cache = {rtype: set() for rtype in _CLEANUP_FUNCS.keys()}
```

And then you `register/unregister` only names in `cache[rtype]` and at the end, you just loop over the different `rtype`/`cleanup` and their respective `cache`.

pierreglaser replied:
Yes, this is definitely something to consider. I was waiting for your input on that topic.

tomMoral commented on loky/backend/spawn.py:
```diff
@@ -83,10 +83,12 @@ def get_preparation_data(name, init_main_module=True):
-        d['tracker_pid'] = semaphore_tracker._semaphore_tracker._pid
+        # TODO: enable resource tracking for other resources than semaphore
+        # on windows?
+        # Pass the resource_tracker pid to avoid re-spawning it in every child
```
Yes, this will be needed for windows platform too I guess.

tomMoral commented on tests/_executor_mixin.py:
```diff
@@ -57,6 +57,7 @@ def _running_children_with_cmdline(p):
     all_children = _direct_children_with_cmdline(p)
     workers = [(c, cmdline) for c, cmdline in all_children
                if (u'semaphore_tracker' not in cmdline and
```
You should also remove the `semaphore_tracker` as it is not here anymore. No?

pierreglaser replied:
I was keeping this line in case a `multiprocessing.sempahore_tracker` was launched for some reason during a test.

tomMoral commented on tests/test_resource_tracker.py:
```diff
@@ -0,0 +1,227 @@
+    return resource_tracker._resource_tracker._pid
+
+
+class TestSemaphoreTracker:
```
```suggestion
class TestResourceTracker:
```

tomMoral commented on tests/test_resource_tracker.py:
```diff
@@ -0,0 +1,227 @@
+class TestSemaphoreTracker:
+
+    @pytest.mark.skipif(sys.platform == "win32",
+                        reason="no resource_tracker on windows")
```
This should not be skipped anymore I guess.

pierreglaser replied:
(cf. TODO) I was waiting for your input on that.

tomMoral commented on tests/test_resource_tracker.py:
```diff
@@ -0,0 +1,227 @@
+
+        from loky import ProcessPoolExecutor
+        from loky.backend import resource_tracker
+        from loky.backend.semlock import SemLock
```
If you do not use the `SemLock` this should be possible to use the same test on `win32` platforms.

pierreglaser commented:
@tomMoral I addressed most of the comments from your previous review.

There is one thing though apparently, `spawn_passfd` (used to create the resource tracker) is POSIX only. For this reason, the resource_tracker tests are failing on Windows.

Recent support  has been added to enable files descriptors to child processes in windows, (post windows 7, and post python3.7). Should we dive further into it? Or stick to POSIX-only support?

tomMoral commented on tests/test_resource_tracker.py:
```diff
@@ -0,0 +1,220 @@
+        _resource_unlink(name1, rtype)
+        p.terminate()
+        p.wait()
+        time.sleep(2.0)
```
Add a comment to say that this timeout is needed to let the `ressource_tracker` the time to end.
To make this test more stable, couldn't you retrieve the semaphore process `pid` and wait is?

pierreglaser replied:
> couldn't you retrieve the semaphore process pid and wait is?

No, we cannot wait on a grand-child process.

tomMoral commented on loky/backend/popen_loky_win32.py:
```diff
@@ -47,9 +48,9 @@ def __init__(self, process_obj):
-        if sys.version_info[:2] > (3, 3):
-            wfd = msvcrt.open_osfhandle(wfd, 0)
+        rfd, wfd = os.pipe()
+        rhandle = duplicate(msvcrt.get_osfhandle(rfd), inheritable=True)
```
Nitpick: use `reduction.duplicate` or `duplicate` everywhere. (see l 99)

tomMoral commented on loky/backend/popen_loky_win32.py:
```diff
@@ -58,10 +59,9 @@ def __init__(self, process_obj):
             with open(wfd, 'wb') as to_child:
                 # start process
                 try:
-                    inherit = sys.version_info[:2] < (3, 4)
```
Just for the sake of understanding, why do you need `inherit=True` for all version?

pierreglaser replied:
**Remainder**: 
Whatever the value of `inherit`, we need to pass the write end of the `resource_tracker` to the child process.

* `inherit=True` means that all `inheritable` handles in the parent process will be usable in the child. Therfore, to pass the write end of the `resource_tracker`, we can simply create an inheritable version of it in the parent, and pass its value to the child in `preparation_data`
* `inherit=False` means that no `handle` can be inherited from the parent to its children. Therefore, to communicate the write end of the `resource_tracker` to its child,  one has to open the parent process handle, snag the pipe handle we want,  and duplicate it - **all of this in the child process**. 

One can argue that the second mechanism is cleaner than the first one. Indeed, in the first case, no parent-process inherited file/pipe handle is  initially available in the child process.


**HOWEVER**:

* this mechanism relies on `winapi` functions like `OpenProcess`, that are only available in `python>=3.4`.
* coincidentally, keep in mind that the case where `inherit=False` used to be implemented natively in `python multiprocessing`, as for `python >= 3.4`, `multiprocessing.spawn.popen_spawn_win32.main` was used as child processes starting command. So the code complexity of this case was abstracted away from us.

Now, `loky.spawn.popen_loky_win32.main` is used for all python versions, to take into account `resource_tracker` information passing at the begining of the process. So if I want to use `inherit=False` (thus steal handles etc.) for `python >= 3.4`, the code will have to be written inside loky. And it is going to result in more version branching.

So I decided to rely only on the first case to pass file handles between a parent and a child, to avoid adding more complexity to the code.

tomMoral replied:
Thanks for the explanation. I think this is fine as is but could you add a comment stating that when we drop `2.7` support, this should be changed?

tomMoral commented on loky/backend/resource_tracker.py:
```diff
@@ -122,36 +143,37 @@ def ensure_running(self):
+        '''Check for the existence of the resource tracker process.'''
         try:
-            self._send('PROBE', '')
+            self._send('PROBE', '', 'noop')
```
Ok I get is now, see comment bellow.

pierreglaser commented on loky/backend/resource_tracker.py:
```diff
@@ -122,36 +143,37 @@ def ensure_running(self):
                         signal.pthread_sigmask(signal.SIG_UNBLOCK,
                                                _IGNORED_SIGNALS)
-            except BaseException:
+            except BaseException as e:
```
Yes we can remove it. This was probably a debugging leftover.

pierreglaser commented on loky/backend/resource_tracker.py:
```diff
@@ -183,32 +205,39 @@ def main(fd, verbose=0):
+                    cmd, rtype, name = line.strip().decode('ascii').split(
+                        ':', 2)
+                    cleanup_func = _CLEANUP_FUNCS.get(rtype, None)
+                    if cleanup_func is None:
```
The way it is written here is how we ended up writting in in python/cpython#13222.

tomMoral replied:
Yes they are equivalent but it is harder to read so I think it would be better to change it. (when I read the code, I look where the `cleanup_func` variable is used)

pierreglaser commented on loky/backend/resource_tracker.py:
```diff
@@ -218,32 +247,33 @@ def main(fd, verbose=0):
+                    except Exception as e:
+                        warnings.warn('resource_tracker: %s: %r' % (name, e))
+                finally:
+                    pass
```
It does not seem useful, let's try to remove it.

pierreglaser commented on loky/backend/resource_tracker.py:
```diff
@@ -122,36 +143,37 @@ def ensure_running(self):
             finally:
-                os.close(r)
+                if os.name == "posix":
+                    os.close(r)
```
Actually, I should also close the handle on windows, but using `CloseHandle`.

pierreglaser commented on loky/backend/resource_tracker.py:
```diff
@@ -122,36 +143,37 @@ def ensure_running(self):
 
-    def _send(self, cmd, name):
-        msg = '{0}:{1}\n'.format(cmd, name).encode('ascii')
+    def _send(self, cmd, name, rtype):
```
Done.

pierreglaser commented:
OK @tomMoral I addressed most of your comments. I am somewhat hesitant about rewriting the `probe` handling because I'd rather be coherent with what we did in python/cpython#13222, but if you feel strongly about this I'll change this code.

ogrisel commented on loky/backend/resource_tracker.py:
```diff
@@ -183,32 +207,42 @@ def main(fd, verbose=0):
+                    if cleanup_func is None:
+                        raise ValueError('Cannot register for automatic '
+                                         'cleanup: unknown resource type {}'
+                                         .format(name, rtype))
```
Please put `list(_CLEANUP_FUNCS)` in the error message.

tomMoral commented:
> I am somewhat hesitant about rewriting the probe handling because I'd rather be coherent with what we did in python/cpython#13222, but if you feel strongly about this I'll change this code.

I understand but we already have some slight differences so I think it would be fine. The question is more whether it is useful or not. I think it allows you to avoid having an empty set `'noop': set()` and this makes more sense to me but I am not feeling this strongly about it so if you do not like it, I am fine with the current implementation.
